futures-util/src/select.rs

```diff
@@ -0,0 +1,38 @@
+//! The `select` macro.
+
+#[macro_export]
+macro_rules! select {
+    ($(
+        $name:ident => $body:expr,
+    )*) => { {
+        $(
+            $crate::assert_unpin(&$name);
```

Comment:
`join!` also moves the value. I think it would be more convenient if the macro works out of the box with `!Unpin` futures by calling the `pin_mut!` macro automatically.

It wouldn't be possible to access the futures after `select()` anymore. But that seems dangerous anyway. If you poll the wrong one, you get a panic.

Reply:
> It wouldn't be possible to access the futures after select() anymore. But that seems dangerous anyway. If you poll the wrong one again, you get a panic.

I think that we strongly want to support this use case-- see my comment in the PR description for how this could work. Every `select` I've ever written aside from timeouts was inside of a loop.

Reply:
I also considered the possibility of having `select_loop!` with the loop baked in, but I think that approach would be too restrictive.